feat(FR-1010): Implement folder explorer

ironAiken2 opened the pull request:
resolves #NNN (FR-MMM)
<!-- replace NNN, MMM with the GitHub issue number and the corresponding Jira issue number. -->

<!--
Please precisely, concisely, and concretely describe what this PR changes, the rationale behind codes,
and how it affects the users and other developers.
-->

**Checklist:** (if applicable)

- [ ] Documentation
- [ ] Minium required manager version
- [ ] Specific setting for review (eg., KB link, endpoint or how to setup)
- [ ] Minimum requirements to check during review
- [ ] Test case(s) to demonstrate the difference of before/after

nowgnuesLee commented on packages/backend.ai-ui/package.json:
```diff
@@ -83,8 +86,11 @@
     "graphql": "^16.10.0",
+    "jotai": "^2.12.2",
     "lucide-react": "^0.484.0",
+    "react": "^19.0.0",
```
react is already in peerdependencies.

nowgnuesLee commented on react/src/components/LegacyFolderExplorer.tsx:
```diff
@@ -42,8 +43,8 @@ const LegacyFolderExplorer: React.FC<LegacyFolderExplorerProps> = ({
   const { t } = useTranslation();
-  const [isWritable, setIsWritable] = useState<boolean>(false);
-  const [isSelected, setIsSelected] = useState<boolean>(false);
+  // const [isWritable, setIsWritable] = useState<boolean>(false);
+  // const [isSelected, setIsSelected] = useState<boolean>(false);
```
could u clean up the comments?

ironAiken2 replied:
These are the comments needed to test the existing backend-ai-folder-explorer.

ironAiken2 reviewed (commented):
Thank you for your review. I've fixed the issue you reported.

The logic error on screen size change is caused by the LegacyFolderExplorer component rendering using a ternary operator per screen size, which causes the component to unmount when the screen size changes. This was fixed by using the Splitter's layout.

ironAiken2 commented:
All just applied. @nowgnuesLee 
1. Fixed changing the path by clicking the breadcrumb or filename to reset existing selected items.
2. The problem was caused by undefined when react-query fetched new data. Fixed by declaring a separate state for cache to show existing data if undefined.

agatha197 reviewed (changes requested):
![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/2HueYSdFvL8pOB5mgrUQ/46d439dc-152e-4313-ac83-d4d079a526ea.png)

The text alignment is inconsistent.

ironAiken2 commented:
![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/lSyr8xXz1wdXALkJKzVx/0cd32e21-168c-42bf-abc4-b79304b61b6f.png)

Just fixed :) @agatha197

nowgnuesLee reviewed (approved):
LGTM

yomybaby commented on react/package.json:
```diff
@@ -119,6 +119,24 @@
       "last 1 safari version"
     ]
   },
+  "pnpm": {
+    "overrides": {
+      "react": "^19.0.0",
+      "react-dom": "^19.0.0",
+      "@types/react": "^19.0.0",
+      "@types/react-dom": "^19.0.0"
+    },
+    "peerDependencyRules": {
+      "ignoreMissing": [
+        "react",
+        "react-dom"
+      ],
+      "allowedVersions": {
+        "react": "19",
+        "react-dom": "19"
+      }
+    }
+  },
```
Should we need this file?? I removed it and try to run. It works properly.

yomybaby commented on packages/backend.ai-ui/src/helper/index.ts:
```diff
@@ -308,6 +308,52 @@ export function divideNumberWithUnits(size1: string, size2: string) {
+ * @param arr - The array to filter.
+ * @returns An array containing only non-empty items.
+ */
+export const filterEmptyItem = <T>(
```
Please remove `filterEmptyItem` and use `filterOutEmpty`

yomybaby commented on packages/backend.ai-ui/src/helper/index.ts:
```diff
@@ -308,6 +308,52 @@ export function divideNumberWithUnits(size1: string, size2: string) {
+  return randStr;
+};
+
+export const maskString = (
+  value = '',
+  maskChar = '*',
+  startFrom = 0,
+  maskLength = 0,
+) => {
+  // clamp mask length
+  maskLength =
+    startFrom + maskLength > value.length ? value.length : maskLength;
+  return (
+    value.substring(0, startFrom) +
+    maskChar.repeat(maskLength) +
+    value.substring(startFrom + maskLength, value.length)
+  );
+};
```
If `maskString` is not used in backend.ai-ui yet, please remove it. Alternatively, consider moving `maskString` from webui to bui.

yomybaby commented on react/src/components/LegacyFolderExplorer.tsx:
```diff
@@ -84,6 +83,9 @@ const LegacyFolderExplorer: React.FC<LegacyFolderExplorerProps> = ({
           ...VFolderNameTitleNodeFragment
+          # ControlItemsFragment is used in FileExplorer
+          ...BAIFileExplorerFragment
+          ...ControlItemsFragment
```
You can remove the `ControlItemsFragment` from this component if it is not directly used here, after PR #4002(merged).

yomybaby commented on react/src/components/LegacyFolderExplorer.tsx:
```diff
@@ -164,26 +170,18 @@ const LegacyFolderExplorer: React.FC<LegacyFolderExplorerProps> = ({
+        <Splitter
+          style={{
+            gap: token.size,
+            maxHeight: 'calc(100vh - 220px)',
+          }}
+          layout={xl ? 'horizontal' : 'vertical'}
+        >
+          <Splitter.Panel resizable={false} max={'60%'}>
             {legacyFolderExplorerPane}
-            {vfolderDescription}
-          </BAIFlex>
-        )
+          </Splitter.Panel>
+          <Splitter.Panel min={'40%'}>{vfolderDescription}</Splitter.Panel>
+        </Splitter>
```
The previous layout using `defaultSize` looks better than the new layout.
```suggestion
        <Splitter
          style={{
            gap: token.size,
            maxHeight: 'calc(100vh - 220px)',
          }}
          layout={xl ? 'horizontal' : 'vertical'}
        >
          <Splitter.Panel resizable={false}>
            {legacyFolderExplorerPane}
          </Splitter.Panel>
          <Splitter.Panel defaultSize={500} min={300}>
            {vfolderDescription}
          </Splitter.Panel>
        </Splitter>
```

yomybaby commented on packages/backend.ai-ui/src/components/baiClient/FileExplorer/BAIFileExplorer.tsx:
```diff
@@ -0,0 +1,334 @@
+          columns={columns}
+          loading={files?.items !== cachedFetchedFiles || isFetching}
+          pagination={false}
+          rowSelection={{
```
It would be better if clicking a row also triggers selection.

yomybaby commented on packages/backend.ai-ui/src/components/baiClient/FileExplorer/ActionItems.tsx:
```diff
@@ -0,0 +1,210 @@
+
+    // TODO: show confirmation modal before upload already existing files
+    // TODO: show progress during upload via using bai-notification
+    const uploadFiles: Array<Promise<tus.Upload>> = _.map(
```
When uploading a folder containing many files and subfolders, the Lit version starts uploading immediately, but the React version causes the browser to freeze.

yomybaby commented on packages/backend.ai-ui/src/components/baiClient/FileExplorer/BAIFileExplorer.tsx:
```diff
@@ -0,0 +1,334 @@
+    }
+  }, [files]);
+
+  const FileExplorerFragment = useFragment(
```
The variable should start with a lowercase letter, and since the return value of `useFragment` is not a fragment or FileExplorer but just a VirtualFolder, please rename the variable to `virtualFolder`, `vFolder`, or `vFolderNode`.

yomybaby commented on packages/backend.ai-ui/src/components/baiClient/FileExplorer/ControlItems.tsx:
```diff
@@ -0,0 +1,154 @@
+import { graphql, useFragment } from 'react-relay';
+
+interface ControlItemsProps {
+  ControlItemsFrgmt?: ControlItemsFragment$key | null;
```
The variable name should indicate its type, so it is clear whether it is a `VirtualFolder`, `ComputeSession`, etc. Please improve the variable name accordingly. Also, props should never start with an uppercase letter.

yomybaby commented on packages/backend.ai-ui/src/components/baiClient/FileExplorer/ControlItems.tsx:
```diff
@@ -0,0 +1,154 @@
+    useMergedAllowedStorageHostPermission();
+  const baiClient = useConnectedBAIClient();
+
+  const currentVFolderHost = useFragment(
```
This is not a VFolderHost but a VFolder, so the variable name should be changed accordingly.

yomybaby reviewed (commented):
LGTM

yomybaby reviewed (approved):
LGTM

nowgnuesLee reviewed (changes requested):
please resolve the conflicts

nowgnuesLee reviewed (approved):
LGTM

agatha197 reviewed (approved):
LGTM

nowgnuesLee reviewed (changes requested):
Please modify it so that the folder size is not visible.

nowgnuesLee commented on packages/backend.ai-ui/src/components/baiClient/FileExplorer/BAIFileExplorer.tsx:
```diff
@@ -0,0 +1,326 @@
+        );
+      },
+    },
+    isDirectorySizeVisible && {
```
When here, the size is not visible even when the type is a file. When it is a directory, please display the size as `-`. However, when `isDirectorySizeVisible` is set, please ensure the size is visible even for directories.

agatha197 commented on packages/backend.ai-ui/src/components/baiClient/FileExplorer/hooks.ts:
```diff
@@ -0,0 +1,57 @@
+import { useQuery } from '@tanstack/react-query';
+import { useState } from 'react';
+
+const directoryTree: Record<string, Array<VFolderFile>> = {};
```
Using a module-level global variable like directoryTree can indeed lead to data conflicts when multiple instances interact with it simultaneously. Moving such variables to a Context or a state management library is a better approach to ensure isolation and predictable behavior.

nowgnuesLee reviewed (approved):
LGTM
